### DISCUSSION TITLE
Enable to obtain WireGuard server public key in one shot (via RPC)

### DISCUSSION THREAD

**Discussion by metalefty** [SoftEtherVPN/SoftEtherVPN#1366] [Ideas]
SoftEtherVPN server administrator needs to share the server's public key (and the pre-shared key) with clients to let clients connect to the server. However, there's no way to obtain the public key in one shot so far.

We can obtain the pre-shared key and the private key by `ProtoOptionsGet` command but need to derive the public key from the private key.

```
VPN Server>protooptionsget wireguard
ProtoOptionsGet command - Lists the options for the specified protocol
Name        |Type   |Value                                       |Description
------------+-------+--------------------------------------------+-----------
Enabled     |Boolean|True                                        |
PresharedKey|String |+m30qGcnKGva12GGIOtzl2vr16C812U10/WU00gOgG4=|
PrivateKey  |String |QdrBf1W/D7uywlsiQBdBGLgFxNU6zk+kyYRHOkysgCo=|
The command completed successfully.
```

```
$ vpncmd /server localhost /csv /cmd:"ProtoOptionsGet WireGuard" | grep PrivateKey | cut -d, -f3 | wg pubkey
ACw6XzmU4su9cz9VpjW9puo5WGmZxXJVu2tYIkNa5SE=
```

This is a little bit inconvenient because an additional tool such as `wireguard-tools` needs to be installed to derive the public key from the private key. It would be very nice if a command like `WgPubkeyGet` to get the public key in one shot.

@davidebeatrici  What do you think?

**Comment by davidebeatrici** [comment]
What about adding a command to obtain the public key from the private one to `vpncmd` instead?

**Reply by metalefty** [reply]
Not bad. That at least satisfy my demand.

**Reply by metalefty** [reply]
Additionally, just an idea, it would be very nice if it is possible to get a sample WireGuard configuration file via `vpncmd` like `OpenVpnMakeConfig` command.

However, I don't think this is necessary because WireGuard configuration is much simpler than OpenVPN. It is possible to write a configuration file from scratch.

```
[Interface]
# Paste your private key 
PrivateKey =
# Set your client's interface IP address
Address =

[Peer]
PublicKey = ${SOFTETHER_WG_PUBKEY}
PreSharedKey = ${SOFETEHER_WG_PRESHAREDKEY}
AllowedIps = ${StaticNetworkGateway}/${SubnetMask} # Add 0.0.0.0/0 if you want to set SoftEtherVPN as default gateway
Endpoint = ${DDNS_HOSTNAME}.softether.net:5555
PersistentKeepAlive = 25
```

**Reply by davidebeatrici** [reply]
In addition to that, we could also add a command to generate a new private key, in order not to depend on WireGuard's tools.

As for the command to create the configuration: it's a good idea, however I believe explaining what SoftEther VPN's options translate to would be better.

**Reply by metalefty** [reply]
Deriving a public key from a private key may be easy. We can just use `crypto_sign_ed25519_sk_to_pk` (secret key to public key). And base64 encode/decode.

**Reply by davidebeatrici** [reply]
Yep!

**Reply by shakibamoshiri** [reply]
> Additionally, just an idea, it would be very nice if it is possible to get a sample WireGuard configuration file via `vpncmd` like `OpenVpnMakeConfig` command.
> 
> However, I don't think this is necessary because WireGuard configuration is much simpler than OpenVPN. It is possible to write a configuration file from scratch.
> 
> ```
> [Interface]
> # Paste your private key 
> PrivateKey =
> # Set your client's interface IP address
> Address =
> 
> [Peer]
> PublicKey = ${SOFTETHER_WG_PUBKEY}
> PreSharedKey = ${SOFETEHER_WG_PRESHAREDKEY}
> AllowedIps = ${StaticNetworkGateway}/${SubnetMask} # Add 0.0.0.0/0 if you want to set SoftEtherVPN as default gateway
> Endpoint = ${DDNS_HOSTNAME}.softether.net:5555
> PersistentKeepAlive = 25
> ```

human/manual intervention increases possibility of more errors. Anything could be automated, it is better be automated.  
With just 80 lines of bash we can auto-generate user.conf file, here is a sample it works fine (tested locally and publicly)

```bash
[Interface]
Address = 192.168.30.1/32
DNS = 192.168.30.1
PrivateKey = sLLZI7gA99JdA7n1nvakUa3xvVpRBkL+K8CDjK7CJlA=

[Peer]
AllowedIPs = 0.0.0.0/0, ::/0
Endpoint = 192.168.1.254:1194
PreSharedKey = 8H4N4NY59KRM4MTQ0w6rMT6PjmF8ke2IIV0ZNQb8afg=
PublicKey = TggPa33QRJ3wAqbQ10GtZtJLa3HUduEuLKAsDUkEDyU= 
```

**Comment by shakibamoshiri** [comment]
it is relatively easy using Bash and a script to automatically generate a "user.conf" file for WireGuard.  
A better one could be using RPC-JSON which currently has not been implemented yet.  
The ideal one could be using "vpncmd <username" to export "username.conf" file

**Comment by lwk523** [comment]
Wish someone can turn this Wireguard into GUI ....
